cmd/kubelet/app/server.go

```diff
@@ -186,6 +186,120 @@ func Run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) error {
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err == nil {
+		clientCertificateManager, err := makeKubeletClientCertificateManager(s, clientConfig, nodeName)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := makeKubeDepsKubeClient(kubeDeps, clientConfig)
+		if err != nil {
+			startClientCertificateManager(kubeClient, clientCertificateManager)
+		}
+
+		setKubeDepsExternalKubeClient(kubeDeps, clientConfig)
+		setKubeDepsEventClient(s, kubeDeps, clientConfig)
+		setKubeDepsHeartbeatClient(s, kubeDeps, clientConfig)
+
+	} else {
+		switch {
+		case s.RequireKubeConfig:
+			return fmt.Errorf("invalid kubeconfig: %v", err)
+		case s.KubeConfig.Provided():
+			glog.Warningf("invalid kubeconfig: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func makeKubeletClientCertificateManager(
+	s *options.KubeletServer,
+	clientConfig *restclient.Config,
+	nodeName types.NodeName) (certificate.Manager, error) {
+
+	var clientCertificateManager certificate.Manager
+	var err error
+	if s.RotateCertificates && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletClientCertificate) {
+		clientCertificateManager, err = kubeletcertificate.NewKubeletClientCertificateManager(
+			s.CertDirectory,
+			nodeName,
+			clientConfig.CertData,
+			clientConfig.KeyData,
+			clientConfig.CertFile,
+			clientConfig.KeyFile)
+		if err != nil {
+			return nil, err
+		}
+		// we set exitIfExpired to true because we use this client configuration to request new certs - if we are unable
+		// to request new certs, we will be unable to continue normal operation
+		if err := kubeletcertificate.UpdateTransport(wait.NeverStop, clientConfig, clientCertificateManager, true); err != nil {
+			return nil, err
+		}
+	}
+
+	return clientCertificateManager, nil
+}
+
+func makeKubeDepsKubeClient(kubeDeps *kubelet.Dependencies, clientConfig *restclient.Config) (clientset.Interface, error) {
+	kubeClient, err := clientset.NewForConfig(clientConfig)
+	if err != nil {
+		glog.Warningf("New kubeClient from clientConfig error: %v", err)
+	}
+
+	kubeDeps.KubeClient = kubeClient
+	return kubeClient, err
+}
+
+func startClientCertificateManager(kubeClient clientset.Interface, clientCertificateManager certificate.Manager) {
+	if kubeClient.CertificatesV1beta1() != nil && clientCertificateManager != nil {
+		glog.V(2).Info("Starting client certificate rotation.")
+		clientCertificateManager.SetCertificateSigningRequestClient(kubeClient.CertificatesV1beta1().CertificateSigningRequests())
+		clientCertificateManager.Start()
+	}
+}
+
+func setKubeDepsExternalKubeClient(kubeDeps *kubelet.Dependencies, clientConfig *restclient.Config) {
+	externalKubeClient, err := clientset.NewForConfig(clientConfig)
+	if err != nil {
+		glog.Warningf("New kubeClient from clientConfig error: %v", err)
+	}
+
+	kubeDeps.ExternalKubeClient = externalKubeClient
+}
+
+func setKubeDepsEventClient(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, clientConfig *restclient.Config) {
```

Comment:
Only pass in the parameters you need from KubeletServer

Reply:
ok